File: works/android/Serializable/app/src/main/java/com/example/serializable/Adapter.java
package com.example.serializable;

import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

public class Adapter extends RecyclerView.Adapter<Adapter.ViewHolder> {
    private final Address[] addresses;

    public Adapter(Address[] addresses) {
        this.addresses = addresses;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        public ConstraintLayout layout;
        public TextView textView;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            this.layout = itemView.findViewById(R.id.recyclerview_item);
            this.textView = itemView.findViewById(R.id.textView);
        }
    }

    @Override @NonNull
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(
                LayoutInflater.from(parent.getContext()).inflate(
                        R.layout.recyclerview_item, parent, false
                )
        );
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        final int adapterPosition = holder.getAdapterPosition();

        holder.textView.setText(addresses[adapterPosition].toString());
        holder.layout.setOnClickListener(view -> {
            Intent intent = new Intent(view.getContext(), NextActivity.class);
            intent.putExtra("Address", addresses[adapterPosition]);
            view.getContext().startActivity(intent);
        });
    }

    @Override
    public int getItemCount() {
        return addresses.length;
    }
}

File: software/RecyclerView/app/src/main/java/com/example/recyclerview/Adapter.kt
package com.example.recyclerview

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

class Adapter(
    private val addresses: List<Address>,
    private val onItemClick: (Int) -> Unit,
    ) : RecyclerView.Adapter<Adapter.ViewHolder>() {

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val textView: TextView = itemView.findViewById(R.id.textView)

        init {
            itemView.setOnLongClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION)
                    onItemClick(adapterPosition)
                return@setOnLongClickListener true
            }
        }

        fun bind(address: Address) = itemView.apply {
            textView.text = address.toString()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.recyclerview_item, parent, false
            )
        )
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        viewHolder.bind(addresses[viewHolder.adapterPosition])
    }

    override fun getItemCount(): Int {
        return addresses.size
    }
}